cpp/src/tsne/bh_kernels.cuh

```diff
@@ -508,21 +507,18 @@ __global__ __launch_bounds__(THREADS4, FACTOR4) void SortKernel(
  */
 __global__ __launch_bounds__(
   THREADS5,
-  FACTOR5) void RepulsionKernel(/* int *restrict errd, */
-                                const float theta,
-                                const float
-                                  epssqd,  // correction for zero distance
-                                const int *restrict sortd,
-                                const int *restrict childd,
-                                const float *restrict massd,
-                                const float *restrict posxd,
-                                const float *restrict posyd,
-                                float *restrict velxd, float *restrict velyd,
-                                float *restrict Z_norm,
-                                const float theta_squared, const int NNODES,
-                                const int FOUR_NNODES, const int N,
-                                const float *restrict radiusd_squared,
-                                const int *restrict maxdepthd) {
+  1) void RepulsionKernel(/* int *restrict errd, */
+                          const float theta,
+                          const float epssqd,  // correction for zero distance
+                          const int *restrict sortd, const int *restrict childd,
+                          const float *restrict massd,
```

Comment:
Now that the KNN is parametrized, the output precision should match `value_t`, rather than assuming it's ` float` everywhere. This caused me to scrape through all the UMAP code and parametrize the functions w/ `value_t` (and `value_idx` where appropriate). 

(And as I'm typing this, I'm looking at a couple places where I missed this on UMAP, so I'll be filing a PR for those shortly).

Reply:
Can you explain what you mean when you say KNN is parametrized? I still am seeing `float` here https://github.com/rapidsai/cuml/pull/3293

Reply:
This is the API for the sparse kNN: https://github.com/rapidsai/cuml/blob/9fa711298080a776210c5fdb77b188606c03e9c6/cpp/src_prims/sparse/knn.cuh#L119. TSNE's API shouldn't be coupled to the KNN API.

Updating the dense API to add the value type param would also be a welcome change if you wanted to cover that here as well. Otherwise, it should get updated when #2821 is worked.

Reply:
I am still a bit confused. Until brute force KNN takes template parameters, TSNE is stuck as well, right? The templates for TSNE won't work until dense and sparse KNN can derive from value_t, unless I'm missing something here

Reply:
The instantiations should be propagating the types all the way through the algorithm. UMAP is a great example where the kNN indices are currently instantiated to int64_t with dense inputs in the user-facing API and with int for sparse inputs. Once #2821 is complete, all we need to do is change the type in the instantiation for dense inputs to int (and adjust the Cython, of course) and the type propagates all the way through.